core/src/storage/collections/binary_heap/impls.rs

```diff
@@ -0,0 +1,448 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+// This file is part of ink!.
+//
+// ink! is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::duplex_sync_chunk::DuplexSyncChunk;
+use crate::storage::{
+    self,
+    alloc::{
+        Allocate,
+        AllocateUsing,
+        Initialize,
+    },
+    chunk::SyncChunk,
+    Flush,
+};
+use core::cmp::{
+    Ord,
+    Ordering,
+};
+#[cfg(feature = "ink-generate-abi")]
+use ink_abi::{
+    HasLayout,
+    LayoutField,
+    LayoutStruct,
+    StorageLayout,
+};
+use scale::{
+    Codec,
+    Decode,
+    Encode,
+};
+#[cfg(feature = "ink-generate-abi")]
+use type_metadata::Metadata;
+
+/// We implement a binary tree.
+pub const CHILDREN: u32 = 2;
+
+/// A binary heap collection.
+/// The heap depends on `Ord` and is a max-heap by default. In order to
+/// make it a min-heap implement the `Ord` trait explicitly on the type
+/// which is stored in the heap.
+///
+/// Provides `O(log(n))` push and pop operations.
+#[derive(Debug)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub struct BinaryHeap<T> {
+    /// The number of nodes stored in the heap.
+    len: storage::Value<u32>,
+    /// The nodes of the heap.
+    entries: DuplexSyncChunk<T>,
+}
+
+/// Iterator over the values of a heap.
+#[derive(Debug)]
+pub struct Values<'a, T> {
+    /// The underlying iterator.
+    iter: Iter<'a, T>,
+}
+
+impl<'a, T> Values<'a, T> {
+    /// Creates a new iterator for the given storage heap.
+    pub(crate) fn new(heap: &'a BinaryHeap<T>) -> Self
+    where
+        T: scale::Codec + Ord,
+    {
+        Self { iter: heap.iter() }
+    }
+}
+
+impl<T> Flush for BinaryHeap<T>
+where
+    T: Ord + Encode + Flush,
+    DuplexSyncChunk<T>: Flush,
+{
+    fn flush(&mut self) {
+        self.len.flush();
+        self.entries.flush();
+    }
+}
+
+#[cfg(feature = "ink-generate-abi")]
+impl<T: Ord> HasLayout for BinaryHeap<T>
+where
+    T: Metadata + 'static,
+{
+    fn layout(&self) -> StorageLayout {
+        LayoutStruct::new(
+            Self::meta_type(),
+            vec![
+                LayoutField::of("len", &self.len),
+                LayoutField::of("entries", &self.entries),
+            ],
+        )
+        .into()
+    }
+}
+
+impl<'a, T> Iterator for Values<'a, T>
+where
+    T: Codec + Ord,
```

Comment:
Please remove all trait bounds that are locally not necessary.

Reply:
:+1: I removed some other bounds, these ones are necessary though because the `Iterator` impl needs access to e.g. `heap.get()` which requires `Ord` when comparing nodes against each other and `Codec` when interacting with the `DuplexSyncChunk` impl (which requires `Codec`).